crates/chain/src/canonical_iter.rs

```diff
@@ -342,3 +385,102 @@ impl<A: Clone> CanonicalReason<A> {
         }
     }
 }
+
+struct TopologicalIter<F>
+where
+    F: FnMut(Txid) -> u32,
+{
+    current_level: Vec<Txid>,
+    next_level: Vec<Txid>,
+
+    adj_list: HashMap<Txid, Vec<Txid>>,
+    inputs_count: HashMap<Txid, usize>,
+
+    current_index: usize,
+
+    sort_by: F,
+}
+
+impl<F> TopologicalIter<F>
+where
+    F: FnMut(bitcoin::Txid) -> u32,
```

Comment:
Would it make it too complicated if we change the signature of this closure to return a sort-key as well as the `Iterator::Item` type?

This will help us reduce potential IO calls later on.